### PR TITLE
AWS SageMaker Hyrbid Pipeline

### PR DESCRIPTION
**Pull request by swiftdiaries**
1. Adds MXNet pipeline for BLERSSI
2. Adds train-upload-to-s3 component
3. Adds inference notebook for model validation from SageMaker Endpoint

### CODE REVIEW

File: apps/networking/ble-localization/hybrid-aws/pipelines/components/v1/mxnet-blerssi-train-upload/mx-dnn.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+bucket_name = 'mxnet-model-store'  # name of the bucket
+s3_client = boto3.client(
+    's3',
+    aws_access_key_id=os.getenv('AWS_ACCESS_KEY_ID'),
+    aws_secret_access_key=os.getenv('AWS_SECRET_ACCESS_KEY'))
+
+response = s3_client.upload_file("model.tar.gz", bucket_name,
+                                    "blerssi/model.tar.gz")
+
```
Comment (ramdootp) [inline apps/networking/ble-localization/hybrid-aws/pipelines/components/v1/mxnet-blerssi-train-upload/mx-dnn.py:124]: the below code section is not needed

File: apps/networking/ble-localization/hybrid-aws/pipelines/components/v1/mxnet-blerssi-train-upload/mx-dnn.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+pred = nd.array([[-200, -200, -200, -75, -200, -200, -200, -200, -200, -200, -200, -200, -200]]) 
+pred_data = gluon.data.DataLoader(gluon.data.ArrayDataset(pred),batch_size=1)
+for i, data in enumerate(pred_data):
+   data = data.as_in_context(model_ctx)
+   output = deserialized_net(data)
+   predictions = nd.argmax(output, axis=1)
+
+print("########################################")
+print("Output, Predictions")
+print(output, predictions)
```
Comment (ramdootp) [inline apps/networking/ble-localization/hybrid-aws/pipelines/components/v1/mxnet-blerssi-train-upload/mx-dnn.py:139]: this section is not needed

File: apps/networking/ble-localization/hybrid-aws/pipelines/blerssi-aws.ipynb [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+    "timestamp = datetime.now().strftime(\"%d-%m-%y-%H-%M-%S\")\n",
+    "logging.info(\"timestamp for pipeline artifacts: %s\", timestamp)\n",
+    "\n",
+    "# pipeline parameters :\n",
+    "\n",
+    "# Region where the pipeline is supposed to push/pull artifacts\n",
+    "aws_region = 'us-west-2'\n",
+    "\n",
+    "# pre-built inference image for serving the mxnet BLERSSI model\n",
+    "inference_image = '609615689023.dkr.ecr.us-west-2.amazonaws.com/mxnet-blerssi-inference:latest'\n",
```
Comment (ramdootp) [inline apps/networking/ble-localization/hybrid-aws/pipelines/blerssi-aws.ipynb:202]: its better to have a pointer to the source  for this image?

**Comment by ramdootp** [comment]
blenet-0100.* files under client dir can be removed

File: apps/networking/ble-localization/hybrid-aws/pipelines/components/v1/mxnet-blerssi-train-upload/mx-dnn.py (hunk 2 above)
Comment (swiftdiaries) [reply apps/networking/ble-localization/hybrid-aws/pipelines/components/v1/mxnet-blerssi-train-upload/mx-dnn.py]: Addressed

File: apps/networking/ble-localization/hybrid-aws/pipelines/blerssi-aws.ipynb (hunk 3 above)
Comment (swiftdiaries) [reply apps/networking/ble-localization/hybrid-aws/pipelines/blerssi-aws.ipynb:211]: Added source code for inference image and linked it in the notebook

File: apps/networking/ble-localization/hybrid-aws/pipelines/README.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    Sample:  "http://10.10.10.10:31380/pipeline"
+
+
+role_arn (string): SageMaker Role ARN for execution of pipeline components
+    Sample: 'arn:aws:iam::${account_id}:role/service-role/AmazonSageMaker-ExecutionRole-${timestemp}'
+```
+
+## Run inference notebook after successful pipeline run
+
+Set the `endpoint_name` variable to the SageMaker endpoint name from a successful pipeline run and step through the notebook to run an inference job.
```
Comment (prem0912) [inline apps/networking/ble-localization/hybrid-aws/pipelines/README.md:64]: Link for predict.ipynb should be provided and mention about predict.ipynb

File: apps/networking/ble-localization/hybrid-aws/pipelines/blerssi-aws.ipynb [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
```
Comment (prem0912) [inline apps/networking/ble-localization/hybrid-aws/pipelines/blerssi-aws.ipynb:211]: Unable to access inference_image from my account, should be made public

File: apps/networking/ble-localization/hybrid-aws/pipelines/components/v1/mxnet-blerssi-train-upload/mx-dnn.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    tarinfo.name = os.path.basename(tarinfo.name)
+    return tarinfo
+
+tar = tarfile.open("model.tar.gz", "w:gz")
+tar.add("model", filter=flatten)
+tar.close()
+
+bucket_name = 'mxnet-model-store'  # name of the bucket
+s3_client = boto3.client(
+    's3',
```
Comment (prem0912) [inline apps/networking/ble-localization/hybrid-aws/pipelines/components/v1/mxnet-blerssi-train-upload/mx-dnn.py:116]: s3 bucket name should be passed through pipeline component. I have tested it by recreating the image

File: apps/networking/ble-localization/hybrid-aws/pipelines/blerssi-aws.ipynb [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+    "\n",
+    "# endpoint config name for the SageMaker Model Serving Endpoint Config\n",
+    "endpoint_config_name = 'mxnet-blerssi-endpoint-config-'+timestamp\n",
+    "\n",
+    "# endpoint name for SageMaker Serving Endpoint\n",
+    "endpoint_name = 'mxnet-blerssi-'+timestamp\n",
+    "\n",
+    "# model artifact URL\n",
+    "# Path to the model tarball \n",
+    "model_path = 's3://mxnet-model-store/blerssi/model.tar.gz'\n",
```
Comment (prem0912) [inline apps/networking/ble-localization/hybrid-aws/pipelines/blerssi-aws.ipynb:224]: its better to pass model path bucket name like role_arn and execute_mode

File: apps/networking/ble-localization/hybrid-aws/pipelines/README.md (hunk 4 above)
Comment (swiftdiaries) [reply apps/networking/ble-localization/hybrid-aws/pipelines/README.md:64]: Merged with pipeline notebook

File: apps/networking/ble-localization/hybrid-aws/pipelines/blerssi-aws.ipynb (hunk 5 above)
Comment (swiftdiaries) [reply apps/networking/ble-localization/hybrid-aws/pipelines/blerssi-aws.ipynb:200]: Set ECR image to be accessible by all AWS users

File: apps/networking/ble-localization/hybrid-aws/pipelines/blerssi-aws.ipynb (hunk 7 above)
Comment (swiftdiaries) [reply apps/networking/ble-localization/hybrid-aws/pipelines/blerssi-aws.ipynb]: Added bucket_name parameter to pass an arbitrary bucket. Requires bucket to be created first.

**Pull request merged**